pkg/ring/replication_set.go

```diff
@@ -10,25 +10,50 @@ import (
 // many errors to tolerate.
 type ReplicationSet struct {
 	Ingesters []IngesterDesc
+
+	// Maximum number of tolerated failing instances. Max errors and max unavailable zones are
+	// mutually exclusive.
 	MaxErrors int
+
+	// Maximum number of different zones in which instances can fail. Max unavailable zones and
+	// max errors are mutually exclusive.
+	MaxUnavailableZones int
 }
 
 // Do function f in parallel for all replicas in the set, erroring is we exceed
 // MaxErrors and returning early otherwise.
 func (r ReplicationSet) Do(ctx context.Context, delay time.Duration, f func(context.Context, *IngesterDesc) (interface{}, error)) ([]interface{}, error) {
+	type instanceResult struct {
```

Comment:
This method would benefit from merging `instanceResult` and `instanceError` and the two channels (`errs` and `resultsChan`). It gets shorter and easier to follow. (See https://gist.github.com/pstibrany/153b237d0dcd861828a706c089ebc8cf)

Reply:
It is indeed shorter and easier to follow. Applied.